File: QuantLibAddin/qlo/bonds.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */

/*
 Copyright (C) 2006, 2007 Chiara Fornarola
 Copyright (C) 2006, 2007, 2008, 2009, 2010, 2011 Ferdinando Ametrano
 Copyright (C) 2005, 2006 Eric Ehlers
 Copyright (C) 2005 Plamen Neykov
 Copyright (C) 2005 Walter Penschke
 Copyright (C) 2009 Piter Dias

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
*/

#if defined(HAVE_CONFIG_H)
    #include <qlo/config.hpp>
#endif

#include <qlo/bonds.hpp>
#include <qlo/flowanalysis.hpp>

#include <ql/instruments/bonds/fixedratebond.hpp>
#include <ql/instruments/bonds/zerocouponbond.hpp>
#include <ql/instruments/bonds/cmsratebond.hpp>
#include <ql/instruments/bonds/floatingratebond.hpp>
#include <ql/cashflows/couponpricer.hpp>
#include <ql/pricingengines/bond/discountingbondengine.hpp>
#include <ql/currency.hpp>
#include <ql/interestrate.hpp>

#include <oh/repository.hpp>

using std::vector;
using std::string;
using boost::shared_ptr;
using ObjectHandler::property_t;
using ObjectHandler::convert2;
using QuantLib::Size;
using QuantLib::Date;

namespace QuantLibAddin {

    vector<vector<property_t> > Bond::flowAnalysis(const Date& d)
    {
        shared_ptr<QuantLib::Bond> temp;
        getLibraryObject(temp);
        const QuantLib::Leg& cashflows = temp->cashflows();

        return QuantLibAddin::flowAnalysis(cashflows, d);
    }

    QuantLib::Real Bond::redemptionAmount() {
        shared_ptr<QuantLib::Bond> temp;
        getLibraryObject(temp);
        return temp->redemption()->amount();
    }

    Date Bond::redemptionDate() {
        shared_ptr<QuantLib::Bond> temp;
        getLibraryObject(temp);
        return temp->redemption()->date();
    }

    string Bond::description() {
        return boost::get<string>(propertyValue("DESCRIPTION"));
    }

    //QuantLib::Currency Bond::currency() {
    //    return boost::get<QuantLib::Currency>(propertyValue("CURRENCY"));
    //}

    string Bond::currency() {
        return boost::get<string>(propertyValue("CURRENCY"));
    }

    void Bond::setCouponPricer(
                const shared_ptr<QuantLib::FloatingRateCouponPricer>& pricer) {
        shared_ptr<QuantLib::Bond> temp;
        getLibraryObject(temp);
        const QuantLib::Leg& cashflows = temp->cashflows();
        QuantLib::setCouponPricer(cashflows, pricer);
    }

    void Bond::setCouponPricers(
            const vector<shared_ptr<QuantLib::FloatingRateCouponPricer> >& p) {
        shared_ptr<QuantLib::Bond> temp;
        getLibraryObject(temp);
        const QuantLib::Leg& cashflows = temp->cashflows();
        QuantLib::setCouponPricers(cashflows, p);
    }

    Bond::Bond(const shared_ptr<ObjectHandler::ValueObject>& properties,
               const string&,
               const QuantLib::Currency&,
               QuantLib::Natural settlementDays,
               const QuantLib::Calendar& calendar,
               QuantLib::Real faceAmount,
               const Date& maturityDate,
               const Date& issueDate,
               const QuantLib::Leg& leg,
               bool permanent) : Instrument(properties, permanent)
    {
        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::Bond(settlementDays,
                           calendar,
                           faceAmount,
                           maturityDate,
                           issueDate,
                           leg));
    }

    ZeroCouponBond::ZeroCouponBond(
            const shared_ptr<ObjectHandler::ValueObject>& properties,
            const string&,
            const QuantLib::Currency&,
            QuantLib::Natural settlementDays,
            const QuantLib::Calendar& calendar,
            QuantLib::Real faceAmount,
            const Date& maturityDate,
            QuantLib::BusinessDayConvention paymentConvention,
            QuantLib::Real redemption,
            const Date& issueDate,
            bool permanent) : Bond(properties, permanent)
    {
        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::ZeroCouponBond(settlementDays,
                                     calendar,
                                     faceAmount,
                                     maturityDate,
                                     paymentConvention,
                                     redemption, issueDate));
    }

    FixedRateBond::FixedRateBond(
            const shared_ptr<ObjectHandler::ValueObject>& properties,
            const string&,
            const QuantLib::Currency&,
            QuantLib::Natural settlementDays,
            QuantLib::Real faceAmount,
            const shared_ptr<QuantLib::Schedule>& schedule,
            const vector<QuantLib::Rate>& coupons,
            const QuantLib::DayCounter& accrualDayCounter,
            QuantLib::BusinessDayConvention paymentConvention,
            QuantLib::Real redemption,
            const Date& issueDate,
            const QuantLib::Calendar& paymentCalendar,
            bool permanent) : Bond(properties, permanent)
    {
        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::FixedRateBond(settlementDays, faceAmount,
                                    *schedule,
                                    coupons, accrualDayCounter,
                                    paymentConvention,
                                    redemption, issueDate,
                                    paymentCalendar));
    }

    FixedRateBond::FixedRateBond(
            const shared_ptr<ObjectHandler::ValueObject>& properties,
            const string&,
            const QuantLib::Currency&,
            QuantLib::Natural settlementDays,
            QuantLib::Real faceAmount,
            const shared_ptr<QuantLib::Schedule>& schedule,
            const vector<shared_ptr<QuantLib::InterestRate> >& coupons,
            QuantLib::BusinessDayConvention paymentConvention,
            QuantLib::Real redemption,
            const Date& issueDate,
            const QuantLib::Calendar& paymentCalendar,
            bool permanent)
    : Bond(properties, permanent)
    {
		vector<QuantLib::InterestRate> couponRate(coupons.size());

		for (Size i=0; i<coupons.size(); ++i)
			couponRate[i] = *coupons[i];

        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::FixedRateBond(settlementDays, faceAmount,
                                    *schedule,
                                    couponRate,
                                    paymentConvention,
									redemption,
                                    issueDate,
                                    paymentCalendar));
    }

    FloatingRateBond::FloatingRateBond(
            const shared_ptr<ObjectHandler::ValueObject>& properties,
            const string&,
            const QuantLib::Currency&,
            QuantLib::Natural settlementDays,
            QuantLib::BusinessDayConvention paymentConvention,
            QuantLib::Real faceAmount,
            const shared_ptr<QuantLib::Schedule>& schedule,
            QuantLib::Natural fixingDays,
            bool inArrears,
            const QuantLib::DayCounter& paymentDayCounter,
            const vector<QuantLib::Rate>& floors,
            const vector<QuantLib::Real>& gearings,
            const shared_ptr<QuantLib::IborIndex>& index,
            const vector<QuantLib::Spread>& spreads,
            const vector<QuantLib::Rate>& caps,
            QuantLib::Real redemption,
            const Date& issueDate,
            bool permanent)
    : Bond(properties, permanent)
    {
        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::FloatingRateBond(settlementDays, faceAmount, *schedule,
                                       index, paymentDayCounter,
                                       paymentConvention, fixingDays,
                                       gearings, spreads,
                                       caps, floors,
                                       inArrears,
                                       redemption, issueDate));
    }

    CmsRateBond::CmsRateBond(
            const shared_ptr<ObjectHandler::ValueObject>& properties,
            const string&,
            const QuantLib::Currency&,
            QuantLib::Natural settlementDays,
            QuantLib::BusinessDayConvention paymentConvention,
            QuantLib::Real faceAmount,
            const shared_ptr<QuantLib::Schedule>& schedule,
            QuantLib::Natural fixingDays,
            bool inArrears,
            const QuantLib::DayCounter& paymentDayCounter,
            const vector<QuantLib::Rate>& floors,
            const vector<QuantLib::Real>& gearings,
            const shared_ptr<QuantLib::SwapIndex>& index,
            const vector<QuantLib::Spread>& spreads,
            const vector<QuantLib::Rate>& caps,
            QuantLib::Real redemption,
            const Date& issueDate,
            bool permanent) : Bond(properties, permanent)
    {
        libraryObject_ = shared_ptr<QuantLib::Instrument>(new
            QuantLib::CmsRateBond(settlementDays, faceAmount, *schedule,
                                  index, paymentDayCounter,
                                  paymentConvention, fixingDays,
                                  gearings, spreads,
                                  caps, floors,
                                  inArrears,
                                  redemption, issueDate));
    }

    vector<string> qlBondAlive(const vector<shared_ptr<Bond> >& bonds,
                               Date& refDate)
    {
        if (bonds.empty())
            return vector<string>();

        if (refDate==Date())
            refDate = QuantLib::Settings::instance().evaluationDate();

        vector<string> result;
        Size n = bonds.size();
        result.reserve(n);

        shared_ptr<QuantLib::Bond> qlBond;
        Date settlement, maturity;
        string id;
        for (Size i=0; i<n; ++i) {
            bonds[i]->getLibraryObject(qlBond);
            settlement = qlBond->settlementDate(refDate);
            maturity = qlBond->maturityDate();
            if (settlement<maturity) {
                id = convert2<string>(bonds[i]->propertyValue("OBJECTID"));
                result.push_back(id);
            }
        }
        return result;
    }

    string qlBondMaturityLookup(const vector<shared_ptr<Bond> >& bonds,
                                const Date& maturity)
    {
        shared_ptr<QuantLib::Bond> qlBond;
        for (Size i=0; i<bonds.size(); ++i) {
            bonds[i]->getLibraryObject(qlBond);
            if (maturity==qlBond->maturityDate())
                return convert2<string>(bonds[i]->propertyValue("OBJECTID"));
        }
        // how to return NA() ?
        return string();
    }

    namespace {

        struct BondItem {
            string objectID;
            Date maturityDate;
            BondItem(const string& _objectID,
                     const Date& _maturityDate)
            : objectID(_objectID),
              maturityDate(_maturityDate) {}
            BondItem() {}
        };

        class BondItemSorter {
          public:
            bool operator()(const BondItem& item1,
                            const BondItem& item2) const
            {
                if (item1.maturityDate>item2.maturityDate)
                    return false;
                return true;
            }
        };
    }

    vector<string> qlBondMaturitySort(const vector<shared_ptr<Bond> >& bonds)
    {
        if (bonds.empty())
            return vector<string>();
        Size n = bonds.size();

        // create BondItems
        std::vector<BondItem> bondItems(n);
        string id;
        Date maturity;
        shared_ptr<QuantLib::Bond> qlBond;
        for (Size i=0; i<n; ++i) {
            bonds[i]->getLibraryObject(qlBond);
            maturity = qlBond->maturityDate() ;
            id = convert2<string>(bonds[i]->propertyValue("OBJECTID"));
            bondItems[i] = BondItem(id, maturity);
        }

        // sort BondItems
        std::sort(bondItems.begin(), bondItems.end(), BondItemSorter());

        // fill result
        std::vector<string> result(n);
        for (Size i=0; i<n; ++i)
            result[i] = bondItems[i].objectID;
        return result;
    }

}

File: QuantLib-SWIG/Python/test/bonds.py
"""
 Copyright (C) 2009 Joseph Malicki

 This file is part of QuantLib, a free-software/open-source library
 for financial quantitative analysts and developers - http://quantlib.org/

 QuantLib is free software: you can redistribute it and/or modify it
 under the terms of the QuantLib license.  You should have received a
 copy of the license along with this program; if not, please email
 <quantlib-dev@lists.sf.net>. The license is also available online at
 <http://quantlib.org/license.shtml>.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the license for more details.
"""

import QuantLib
import unittest

class FixedRateBondTest(unittest.TestCase):
    def setUp(self):
        self.settlement_days = 3
        self.face_amount = 100.0
        self.redemption = 100.0
        self.issue_date = QuantLib.Date(2,1,2008)
        self.maturity_date = QuantLib.Date(2,1,2018)
        self.calendar = QuantLib.UnitedStates(QuantLib.UnitedStates.GovernmentBond)
        self.day_counter = QuantLib.ActualActual(QuantLib.ActualActual.Bond)
        self.sched = QuantLib.Schedule(self.issue_date, self.maturity_date,
                                       QuantLib.Period(QuantLib.Semiannual), self.calendar,
                                       QuantLib.Unadjusted, QuantLib.Unadjusted,
                                       QuantLib.DateGeneration.Backward, False)
        self.coupons = [0.05]

        self.bond = QuantLib.FixedRateBond(self.settlement_days, self.face_amount,
                                           self.sched, self.coupons, self.day_counter,
                                           QuantLib.Following, self.redemption,
                                           self.issue_date)

        self.flat_forward = QuantLib.FlatForward(self.issue_date+self.settlement_days*QuantLib.Days,
                                            self.coupons[0], self.day_counter,
                                            QuantLib.Compounded, QuantLib.Semiannual)
        self.term_structure_handle = QuantLib.RelinkableYieldTermStructureHandle(self.flat_forward)
        bondEngine = QuantLib.DiscountingBondEngine(self.term_structure_handle)
        self.bond.setPricingEngine(bondEngine)

    def testFrequency(self):
        """ Testing FixedRateBond frequency() method. """
        self.assertEqual(self.bond.frequency(), QuantLib.Semiannual)

    def testDayCounter(self):
        """ Testing FixedRateBond dayCounter() method. """
        self.assertEqual(self.bond.dayCounter(), self.day_counter)

    def testSimpleInspectors(self):
        """ Testing FixedRateBond simple inspectors. """
        self.assertEqual(self.bond.settlementDays(), self.settlement_days)
        self.assertEqual(self.bond.notional(), self.face_amount)
        self.assertEqual(self.bond.issueDate(), self.issue_date)
        self.assertEqual(self.bond.maturityDate(), self.maturity_date)

    #def testSettlementValue(self):
    #    """ Testing FixedRateBond settlement value. """
    #    orig_date = QuantLib.Settings.evaluationDate
    #    QuantLib.Settings.evaluationDate = self.issue_date + 1*QuantLib.Months
    #    self.assertEqual(round(self.bond.settlementValue(100.0), 4), 102.3098)
    #    QuantLib.Settings.evaluationDate = orig_date

    def testCashFlows(self):
        """ Testing that the FixedRateBond gives the expected cash flows. """
        self.assertEqual([round(cf.amount(), 4) for cf in self.bond.cashflows()],
                         20*[round(self.face_amount * self.coupons[0] / 2, 4)] + \
                         [round(self.redemption, 4)])

    def testRedemption(self):
        """ Testing FixedRateBond redemption value and date. """
        self.assertEqual(self.bond.redemption().date(), self.maturity_date)
        self.assertEqual(self.bond.redemption().amount(), self.redemption)

    def testRedemptions(self):
        """ Testing FixedRateBond redemptions. """
        redemptions = self.bond.redemptions()
        self.assertEqual(len(redemptions), 1)
        self.assertEqual(redemptions[0].date(), self.maturity_date)
        self.assertEqual(redemptions[0].amount(), self.redemption)

    def testNotional(self):
        """ Testing FixedRateBond notional values. """
        self.assertEqual(self.bond.notional(), 100.0)
        self.assertEqual(self.bond.notionals(), (100.0, 0))

    def testNextCoupon(self):
        """ Testing FixedRateBond correct next coupon amount. """
        self.assertEqual(self.bond.nextCouponRate(self.issue_date), 0.05)

    def testPrevCoupon(self):
        """ Testing FixedRateBond correct previous coupon amount. """
        self.assertEqual(self.bond.previousCouponRate(self.issue_date), 0.05)

    def testCleanPrice(self):
        """ Testing FixedRateBond clean price. """
        self.assertEqual(round(self.bond.cleanPrice(0.05, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days), 4),
                         100.0)
        self.assertEqual(round(self.bond.cleanPrice(0.05, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days +
                                                    1*QuantLib.Months), 4),
                         99.9997)

        self.assertEqual(round(self.bond.cleanPrice(0.06, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days +
                                                    1*QuantLib.Months), 4),
                         92.5639)


    def testDirtyPrice(self):
        """ Testing FixedRateBond dirty price. """
        self.assertEqual(round(self.bond.dirtyPrice(0.05, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days), 4),
                         100.0)
        self.assertEqual(round(self.bond.dirtyPrice(0.05, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days +
                                                    1*QuantLib.Months), 4),
                         100.0271)
        self.assertEqual(round(self.bond.dirtyPrice(0.06, self.day_counter, QuantLib.Compounded,
                                                    QuantLib.Semiannual, self.issue_date +
                                                    self.settlement_days*QuantLib.Days +
                                                    1*QuantLib.Months), 4),
                         92.5913)

    def testCleanPriceFromZSpread(self):
        """ Testing FixedRateBond clean price derived from Z-spread. """
        self.assertEqual(round(QuantLib.cleanPriceFromZSpread(
                    self.bond, self.flat_forward, 0.01,
                    self.day_counter, QuantLib.Compounded, QuantLib.Semiannual,
                    self.issue_date + 1 * QuantLib.Months), 4), 92.5637)

if __name__ == '__main__':
    print 'testing QuantLib', QuantLib.__version__
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(FixedRateBondTest,'test'))
    unittest.TextTestRunner(verbosity=2).run(suite)
    raw_input('press return to continue')
